book/objects.tex

```diff
@@ -844,11 +844,11 @@ \subsection{The Pixel Class}
     has %.color is rw;
 
     method change_color(%hue) {
-        self!color = %hue
+        %!color = %hue
```

Comment:
self!color is correct, and %!color is also correct.

Reply:
As far as i can say, self!color is not a private method invocation, but a direct access to the current object's color attribute.

Reply:
When I run the code (https://glot.io/snippets/f1pszh6vc4), I obtain this error:
`No such private method '!color' for invocant of type 'Pixel'.`

This leads me to believe that perl6 is expecting a private method `method !color {}` when
`!color` is invoked on `self`. And according to the [doc](https://docs.perl6.org/language/objects#self), rather than being a direct access to the current object, `self!color` is a method invocation on the invocant.

Reply:
Oops, you're right. The code that I actually tested was as follows:

class Pixel is Point2D {
    has %.color is rw;

    multi method change_color(%hue) {
        self.color = %hue
    }
    multi method change_color(Int $red, Int $green, Int $blue) {
        # signature using positional parameters
        self.color = (red => $red, green => $green, blue => $blue)
    }
}

And, BTW, this is also what I have in the code in the supplementary section: https://github.com/LaurentRosenfeld/thinkperl6/tree/master/Supplementary.

I have no idea how the dot was changed to an exclamation mark in the book's text.